### PR TITLE
Migrate to Coil version 2.1.0

### PR DESCRIPTION
**Pull request by skydoves**
### 🎯 Goal

Migrate to Coil version [2.1.0](https://github.com/coil-kt/coil/releases/tag/2.1.0).

I referenced the documentation below:
- [Coil CHANGELOG](https://github.com/coil-kt/coil/blob/main/CHANGELOG.md#210---may-17-2022)
- [Official docs](https://coil-kt.github.io/coil/)

### 🛠 Implementation details

- There are API changes for Compose: `rememberAsyncImagePainter`, `AsyncImage`.

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] PR targets the `develop` branch

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

gif

### CODE REVIEW

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/coil/StreamImageLoaderFactory.kt [hunk 1]
```diff
@@ -22,43 +22,36 @@ import coil.ImageLoader
...
-import okhttp3.Dispatcher
-import okhttp3.Interceptor
-import okhttp3.OkHttpClient
+import coil.decode.VideoFrameDecoder
+import coil.disk.DiskCache
+import coil.memory.MemoryCache
 
 private const val DEFAULT_MEMORY_PERCENTAGE = 0.25
+private const val DEFAULT_DISK_CACHE_PERCENTAGE = 0.02
+private const val DISK_CACHE_DIRECTORY = "image_cache"
```
Comment (skydoves) [inline stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/coil/StreamImageLoaderFactory.kt:31]: For here, I referenced the official guide - https://coil-kt.github.io/coil/image_loaders/#caching

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/coil/StreamImageLoaderFactory.kt [hunk 2]
```diff
@@ -22,43 +22,36 @@ import coil.ImageLoader
...
-                val dispatcher = Dispatcher().apply { maxRequestsPerHost = maxRequests }
-
-                OkHttpClient.Builder()
-                    .cache(CoilUtils.createDefaultCache(context))
-                    .dispatcher(dispatcher)
-                    .addNetworkInterceptor(cacheControlInterceptor)
+            .diskCache {
+                DiskCache.Builder()
+                    .directory(context.cacheDir.resolve(DISK_CACHE_DIRECTORY))
+                    .maxSizePercent(DEFAULT_DISK_CACHE_PERCENTAGE)
```
Comment (skydoves) [inline stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/coil/StreamImageLoaderFactory.kt:45]: Coil 1.x relied on OkHttp's disk cache. Coil 2.x has its own disk cache and should not use OkHttp's Cache.

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/images/CoilStreamImageLoader.kt [hunk 3]
```diff
@@ -26,8 +26,7 @@ import android.os.Build
 import android.widget.ImageView
 import coil.drawable.MovieDrawable
 import coil.drawable.ScaleDrawable
-import coil.fetch.VideoFrameUriFetcher
-import coil.loadAny
+import coil.load
```
Comment (skydoves) [inline stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/images/CoilStreamImageLoader.kt:29]: So they renamed `loadAny` to `load`, which forces us to swap the names. 😂

File: docusaurus/docs/Android/03-ui/02-general-customization/02-chatui.mdx [hunk 4]
```diff
@@ -149,11 +149,6 @@ ChatUI.avatarBitmapFactory = object : AvatarBitmapFactory(context) {
         val imageResult = context.imageLoader.execute(
             ImageRequest.Builder(context)
                 .data(user.image)
-                .apply {
-                    if (!user.online) {
-                        transformations(BlurTransformation(context))
-                    }
-                }
```
Comment (skydoves) [inline docusaurus/docs/Android/03-ui/02-general-customization/02-chatui.mdx:156]: `BlurTransformation` has been permanently deprecated.

Comment (skydoves) [reply docusaurus/docs/Android/03-ui/02-general-customization/02-chatui.mdx:156]: What do you think about guiding copy & paste the [BlurTransformation](https://github.com/coil-kt/coil/blob/845f39383f332428077c666e3567b954675ce248/coil-base/src/main/java/coil/transform/BlurTransformation.kt) to users who use Coil 2.x?

From the Coil docs:
> [BlurTransformation](https://github.com/coil-kt/coil/blob/845f39383f332428077c666e3567b954675ce248/coil-base/src/main/java/coil/transform/BlurTransformation.kt) and [GrayscaleTransformation](https://github.com/coil-kt/coil/blob/845f39383f332428077c666e3567b954675ce248/coil-base/src/main/java/coil/transform/GrayscaleTransformation.kt) are removed from the library. If you use them, you can copy their code into your project.

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/images/CoilStreamImageLoader.kt (hunk 3 above)
Comment (skydoves) [reply stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/images/CoilStreamImageLoader.kt:29]: Done: [585cc0d](https://github.com/GetStream/stream-chat-android/pull/3538/commits/585cc0d4bf017cdf797ddc7f433a2108c2434cba)

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/images/StreamImageLoader.kt [hunk 5]
```diff
@@ -35,7 +35,7 @@ public sealed interface StreamImageLoader {
     public var imageHeadersProvider: ImageHeadersProvider
 
     @Suppress("LongParameterList")
-    public fun load(
```
[the remark this answers is not in the document]
Comment (skydoves) [reply stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/images/StreamImageLoader.kt]: Done: [585cc0d](https://github.com/GetStream/stream-chat-android/pull/3538/commits/585cc0d4bf017cdf797ddc7f433a2108c2434cba)

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/images/ViewExtensions.kt [hunk 6]
```diff
@@ -25,14 +25,14 @@ import com.getstream.sdk.chat.images.StreamImageLoader.ImageTransformation
 import io.getstream.chat.android.core.internal.InternalStreamChatApi
 
 @InternalStreamChatApi
-public fun ImageView.load(
```
[the remark this answers is not in the document]
Comment (skydoves) [reply stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/images/ViewExtensions.kt]: Done: [63e756d](https://github.com/GetStream/stream-chat-android/pull/3538/commits/63e756d99aa3a8959c6f71aab013e7828140819f)

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/common/internal/AvatarFetcher.kt [hunk 7]
```diff
@@ -1,72 +0,0 @@
...
-            ),
-            false,
-            DataSource.MEMORY
-        )
-    }
-
-    override fun key(data: Avatar): String? = when (data) {
-        is Avatar.UserAvatar -> ChatUI.avatarBitmapFactory.userBitmapKey(data.user)
-        is Avatar.ChannelAvatar -> ChatUI.avatarBitmapFactory.channelBitmapKey(data.channel)
-    }
```
[the remark this answers is not in the document]
Comment (skydoves) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/common/internal/AvatarFetcher.kt:71]: Oh, I missed the `Keyer`. Added: [5a6e1ee](https://github.com/GetStream/stream-chat-android/pull/3538/commits/5a6e1eed2c4aa7d881e57b4f4a7c2f2b87d6136d)

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/coil/StreamImageLoaderFactory.kt (hunk 2 above)
Comment (skydoves) [reply stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/coil/StreamImageLoaderFactory.kt:65]: Done: [6c3e7fa](https://github.com/GetStream/stream-chat-android/pull/3538/commits/6c3e7fa1e9dcb513d84867c78f3a3a6091b71a3a)

File: docusaurus/docs/Android/03-ui/02-general-customization/02-chatui.mdx (hunk 4 above)
Comment (skydoves) [reply docusaurus/docs/Android/03-ui/02-general-customization/02-chatui.mdx]: Updated with [GrayScaleTransformation](https://github.com/GetStream/stream-chat-android/pull/3538/commits/bba2c8897878d81f3e7f033856a84bd6398f984e). docs: [9d1d434](https://github.com/GetStream/stream-chat-android/pull/3538/commits/9d1d4342835b2841efd99f3db30ad1d3e90b2145).

**Review by JcMinarro — APPROVED** [review]
LGTM

**Review by MarinTolic — APPROVED** [review]
### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/coil/StreamImageLoaderFactory.kt [hunk 8]
```diff
@@ -48,17 +53,23 @@ public class StreamImageLoaderFactory(
                 val dispatcher = Dispatcher().apply { maxRequestsPerHost = maxRequests }
```
Comment (leandroBorgesFerreira) [inline stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/coil/StreamImageLoaderFactory.kt:53]: we don't need `val cacheControlInterceptor` anymore. Or do we?

Comment (filbabic) [reply stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/coil/StreamImageLoaderFactory.kt:53]: The disk and memory cache are now done through two separate functions, so no longer is an interceptor required. (from what I understand)

**Pull request merged**